### PR TITLE
Use protocol provided buffer for requests (fixes #1157)

### LINKED ISSUE #1157: Reduce memory/CPU footprint
Use `sync.Pool` in `model.Request`, because judging from:
https://discourse.syncthing.net/t/disable-data-encryption/416/44
we seem to be making a lot of allocations while sending the 5GB file which seem to be taking a fair share of CPU time.

### PR DESCRIPTION
**Pull request by AudriusButkevicius**
No benchmarks though.

### CODE REVIEW

**Comment by AudriusButkevicius** [comment]
@st-jenkins run integration tests please

**Comment by AudriusButkevicius** [comment]
@st-jenkins run integration tests please

**Comment by AudriusButkevicius** [comment]
@st-jenkins run integration tests please

**Comment by AudriusButkevicius** [comment]
@st-jenkins run integration tests please

File: internal/beacon/beacon.go [hunk 1]
```diff
@@ -11,15 +11,17 @@ import "net"
 type recv struct {
 	data []byte
 	src  net.Addr
+	done chan struct{}
 }
 
 type Interface interface {
 	Send(data []byte)
-	Recv() ([]byte, net.Addr)
+	Recv() ([]byte, net.Addr, chan struct{})
```
Comment (calmh) [inline internal/beacon/beacon.go]: Does the beacon changes really buy us very much, given that we're looking at like one small buffer every thirty seconds or so?

**Comment by calmh** [comment]
What's the advantage of the `bytepool` package over a plain `sync.Pool`?

**Comment by AudriusButkevicius** [comment]
Not a lot, it deals with buffers of a specific size.

**Comment by AudriusButkevicius** [comment]
@st-jenkins run integration tests please

**Comment by Zillode** [comment]
Before: (compares http://build.syncthing.net/view/Testing/job/syncthing-pr/804/console to http://build.syncthing.net/job/syncthing-pr/805/console)
```
Benchmark10kReplace                20     60388016 ns/op
Benchmark10kUpdateChg               5    306595050 ns/op
Benchmark10kUpdateSme              20     88851785 ns/op
Benchmark10kNeed2k                 30     38797488 ns/op
Benchmark10kHaveFullList          100     14241885 ns/op
Benchmark10kGlobal                 10    100060432 ns/op
ok  	github.com/syncthing/syncthing/internal/db	12.418s

BenchmarkMatch              50000        36224 ns/op
BenchmarkMatchCached     10000000          189 ns/op
ok  	github.com/syncthing/syncthing/internal/ignore	4.293s

BenchmarkIndex_10000                      100     17342435 ns/op      2734011 B/op        70135 allocs/op
BenchmarkIndex_100                       2000       813894 ns/op        35867 B/op          821 allocs/op
BenchmarkIndexUpdate_10000_10000           50     31121292 ns/op      4643115 B/op        90139 allocs/op
BenchmarkIndexUpdate_10000_100           1000      1401426 ns/op        52531 B/op         1009 allocs/op
BenchmarkIndexUpdate_10000_1             1000      1296265 ns/op         8068 B/op          117 allocs/op
BenchmarkRequest                     20000000         94.8 ns/op
BenchmarkTree_10000_50                      5    248587136 ns/op    122907361 B/op       624169 allocs/op
BenchmarkTree_100_50                      200      8638800 ns/op      1094204 B/op         5859 allocs/op
BenchmarkTree_100_10                      200      7568059 ns/op       253728 B/op         1853 allocs/op
BenchmarkJobQueueBump                   30000       149529 ns/op
BenchmarkJobQueuePushPopDone10k           500      3507075 ns/op
ok  	github.com/syncthing/syncthing/internal/model	27.279s
```
Pr:
```
Benchmark10kReplace                20     61793258 ns/op
Benchmark10kUpdateChg               5    309457152 ns/op
Benchmark10kUpdateSme              20     89202933 ns/op
Benchmark10kNeed2k                 30     39353746 ns/op
Benchmark10kHaveFullList          100     14162783 ns/op
Benchmark10kGlobal                 10    100099067 ns/op
ok  	github.com/syncthing/syncthing/internal/db	12.498s

BenchmarkMatch              50000        36055 ns/op
BenchmarkMatchCached     10000000          194 ns/op
ok  	github.com/syncthing/syncthing/internal/ignore	4.337s

BenchmarkIndex_10000                      100     17591036 ns/op      2734011 B/op        70135 allocs/op
BenchmarkIndex_100                       2000       831170 ns/op        35630 B/op          821 allocs/op
BenchmarkIndexUpdate_10000_10000           50     32324693 ns/op      4643097 B/op        90139 allocs/op
BenchmarkIndexUpdate_10000_100           1000      1396746 ns/op        52530 B/op         1009 allocs/op
BenchmarkIndexUpdate_10000_1             1000      1288470 ns/op         8068 B/op          117 allocs/op
BenchmarkRequest                     20000000         94.7 ns/op
BenchmarkTree_10000_50                      5    246361569 ns/op    122910184 B/op       624179 allocs/op
BenchmarkTree_100_50                      200      8912803 ns/op      1094195 B/op         5859 allocs/op
BenchmarkTree_100_10                      200      8011735 ns/op       253740 B/op         1853 allocs/op
BenchmarkJobQueueBump                   30000       144459 ns/op
BenchmarkJobQueuePushPopDone10k           500      3471778 ns/op
ok  	github.com/syncthing/syncthing/internal/model	27.367s
```

**Comment by AudriusButkevicius** [comment]
These don't test the part I am changing sensibly. A good benchmark would be a graph of memory usage while transferring a 5GB file which perhaps can be done via: https://github.com/syncthing/syncthing/blob/master/test/transfer-bench_test.go

File: internal/model/model.go [hunk 2]
```diff
@@ -705,19 +705,19 @@ func (m *Model) Close(device protocol.DeviceID, err error) {
 
 // Request returns the specified data segment by reading it from local disk.
 // Implements the protocol.Model interface.
-func (m *Model) Request(deviceID protocol.DeviceID, folder, name string, offset int64, size int, hash []byte, flags uint32, options []protocol.Option) ([]byte, error) {
+func (m *Model) Request(deviceID protocol.DeviceID, folder, name string, offset int64, size int, hash []byte, flags uint32, options []protocol.Option, buf []byte) error {
 	if offset < 0 || size < 0 {
```
Comment (calmh) [inline internal/model/model.go]: Just for fun, `if len(buf) < size { panic("dude...") }`?

Comment (calmh) [inline internal/model/model.go]: Actually, make that check `==` as `ReadAt()` uses the buffer length for the read. Or maybe remove the `size` parameter altogether?

Comment (AudriusButkevicius) [inline internal/model/model.go]: Sorry, but why? What are we scared of here?

Comment (calmh) [inline internal/model/model.go]: Passing a buffer with a len different from the `size` parameter, which is a silent weird failure. The two *must* be the same, so either we ought to enforce that, or remove `size` so there's no possibility of a mismatch?

Comment (AudriusButkevicius) [inline internal/model/model.go]: Ok I'll remove size all together.

**Comment by calmh** [comment]
@st-jenkins run integration tests please

**Comment by calmh** [comment]
Lgtm, just waiting for integration

**Comment by AudriusButkevicius** [comment]
Did you run any benchmakrs by any chanse?

**Comment by calmh** [comment]
I did now. No real difference that I could tell...

**Pull request merged**